### ISSUE TITLE
Unable to determine sheetname

### ISSUE THREAD

**Issue by ct3inc** [mbleron/ExcelTable#11]
I don't know the sheet name of the Excel file so I need to get it dynamically.  I have tried the following but it returns no results:
    SELECT *
      FROM exceltable.getrows(p_blob_data, '.*',
                               '"C1" VARCHAR2(50),
                               "SHEET_NAME" varchar2(31) for metadata (sheet_name)')
   WHERE SHEET_IDX = 1;

    SELECT *
      FROM exceltable.getrows(p_blob_data, '.*',
                               '"C1" VARCHAR2(50),
                               "SHEET_NAME" varchar2(31) for metadata (sheet_name)');

This one will return the sheet name:
  CURSOR c_xls_getsheet1 (p_blob_data BLOB)
  IS
    SELECT *
      FROM exceltable.getrows(p_blob_data, '.*',
                               '"C1" VARCHAR2(50),
                               "SHEET_NAME" varchar2(31) for metadata (sheet_name)',
                               '4:4');

If I have multiple sheets, I haven't found a way to get the sheet name successfully.

**Comment by mbleron** [comment]
> SELECT *
> FROM exceltable.getrows(p_blob_data, '.*',
> '"C1" VARCHAR2(50),
> "SHEET_NAME" varchar2(31) for metadata (sheet_name)')
> WHERE SHEET_IDX = 1;

That one shouldn't even compile since there's no TABLE operator, nor SHEET_IDX column in the list.
Please post the **real** queries you've tried, otherwise I won't be able to see what you're doing wrong.

In order to use the "sheet name as a regex pattern" feature, make sure you have enabled it in the first place via a call to useSheetPattern : 
```
begin
  ExcelTable.useSheetPattern(true);
end;
/
```
or by setting it by default to true in the package body.

Then, you can do this, for example using this [file](https://github.com/mbleron/ExcelTable/blob/master/samples/multisheet.xlsx) : 
```
SQL> select *
  2  from table(
  3         ExcelTable.getRows(
  4           ExcelTable.getFile('XL_DATA_DIR','multisheet.xlsx')
  5         , '.*'
  6         , q'{
  7              "C1"         number column 'A'
  8            , "SHEET_IDX"  number for metadata (sheet_index)
  9            , "SHEET_NAME" varchar2(31 char) for metadata (sheet_name)
 10            }'
 11         )
 12       ) x
 13  ;
 
        C1  SHEET_IDX SHEET_NAME
---------- ---------- --------------------------------------------------------------------------------
         1          1 Sheet2
         2          1 Sheet2
         3          1 Sheet2
         4          2 Sheet1
         5          2 Sheet1
         6          2 Sheet1
         7          3 Sheet3
         8          3 Sheet3
         9          3 Sheet3
 
9 rows selected.
```

If you still have issues, please attach (drag&drop) a sample file here so that I can reproduce your problem.
Thanks.

**Comment by ct3inc** [comment]
Sorry that was a typo on the SHEET_IDX.  I did have the following query:

    SELECT *

      FROM exceltable.getrows(p_blob_data, '.*', 

                              '"C1" VARCHAR2(50),

                               "SHEET_NAME" varchar2(31) for metadata (sheet_name),

                               "SHEET_IDX" NUMBER FOR METADATA (SHEET_INDEX)

      WHERE SHEET_IDX = 1;

 

I did run the following before trying to get the sheet name:

 
begin
  ExcelTable.useSheetPattern(true);
end;
/

 

I’m unable to get the sheet name for the two files.  

 

If I add a range of ‘4:4’ for we90220.xlsx, I’m able to get the sheet name.  Don’t know why this is. 

 

Also, if I try to just get the sheet name from any Excel file it doesn’t work.

 

    SELECT *

      FROM exceltable.getrows(p_blob_data, '.*',

                              '"SHEET_NAME" varchar2(31) for metadata (sheet_name)');

 

Thank you,

 

Mike

**Comment by mbleron** [comment]
Could you post a sample file?
You can mail it to me directly if you prefer (mbleron2 [at] gmail [dot] com).

**Comment by ct3inc** [comment]
[SB_GDRC 5-6-19.xlsx](https://github.com/mbleron/ExcelTable/files/3158600/SB_GDRC.5-6-19.xlsx)
[we90220.xlsx](https://github.com/mbleron/ExcelTable/files/3158603/we90220.xlsx)

Here are the two files that I'm having issues with.

**Comment by mbleron** [comment]
I can't reproduce the issue, see test cases below : 

File 1 : 
```
SQL> select *
  2  from table(
  3         ExcelTable.getRows(
  4           ExcelTable.getFile('TMP_DIR','SB_GDRC.5-6-19.xlsx')
  5         , '.*'
  6         , q'{
  7              "C1"          varchar2(50)
  8            , "SHEET_NAME"  varchar2(31 char) for metadata (sheet_name)
  9            }'
 10         )
 11       ) x
 12  ;
 
C1        SHEET_NAME
--------- -------------
SBC #     GDRC
230389    GDRC
230377    GDRC
230373    GDRC
230372    GDRC
 
```
Note : the 2nd sheet does not appear in the result because there's no data for the first column.

File 2 : 
```
SQL> select *
  2  from table(
  3         ExcelTable.getRows(
  4           ExcelTable.getFile('TMP_DIR','we90220.xlsx')
  5         , '.*'
  6         , q'{
  7              "C1"          varchar2(50)
  8            , "SHEET_NAME"  varchar2(31 char) for metadata (sheet_name)
  9            }'
 10         )
 11       ) x
 12  ;
 
C1    SHEET_NAME
----- --------------
1     we90220
2     we90220
 
```
```
SQL> select *
  2  from table(
  3         ExcelTable.getRows(
  4           ExcelTable.getFile('TMP_DIR','we90220.xlsx')
  5         , '.*'
  6         , q'{
  7              "Ticket No"      number  column 'B'
  8            , "Date for Pay"   date    column 'F'
  9            , "SHEET_IDX"      number            for metadata (sheet_index)
 10            , "SHEET_NAME"     varchar2(31 char) for metadata (sheet_name)
 11            }'
 12         , 'A2'
 13         )
 14       ) x
 15  ;
 
 Ticket No Date for Pay  SHEET_IDX SHEET_NAME
---------- ------------ ---------- -------------
   1314241 20/02/2019            1 we90220
   1314242 20/02/2019            1 we90220
 
```
  

> Also, if I try to just get the sheet name from any Excel file it doesn’t work.  

Indeed, if no data column is selected, it returns an empty result.
I'll see if I can provide a fix.

**Comment by ct3inc** [comment]
That’s weird.  I’m trying to run this in a package and I have a cursor setup as follows:

 

  CURSOR c_we_getrows (p_blob_data BLOB,

                       p_sheet_name VARCHAR2)

  IS

    SELECT *

      FROM exceltable.getrows(p_file  => p_blob_data, 

                              p_sheet => p_sheet_name,

                              p_cols  => q'{"A"  VARCHAR2(30) column 'A',

                                            "B"  VARCHAR2(30) column 'B',

                                            "C"  VARCHAR2(30) column 'C',

                                            "D"  VARCHAR2(30) COLUMN 'D',

                                            "E"  VARCHAR2(30) column 'E',

                                            "F"  VARCHAR2(30) COLUMN 'F',

                                            "G"  VARCHAR2(30) column 'G',

                                            "H"  VARCHAR2(30) COLUMN 'H',

                                            "I"  VARCHAR2(30) column 'I',

                                            "J"  VARCHAR2(30) COLUMN 'J',

                                            "K"  VARCHAR2(30) column 'K',

                                            "L"  VARCHAR2(30) COLUMN 'L',

                                            "M"  VARCHAR2(30) column 'M',

                                            "N"  VARCHAR2(30) COLUMN 'N',

                                            "O"  VARCHAR2(30) column 'O',

                                            "P"  VARCHAR2(30) COLUMN 'P',

                                            "Q"  VARCHAR2(30) column 'Q',

                                            "R"  VARCHAR2(30) COLUMN 'R'}',

                              p_range => '1:500');

  r_we_getrows        c_we_getrows%ROWTYPE;

 

But I can’t compile because I get the following error:  Error(449,5): PL/SQL: ORA-06550: line 4, column 13: PLS-00307: too many declarations of 'ODCITABLEDESCRIBE' match this call ORA-06550: line 4, column 6: PL/SQL: Statement ignored 

If I disable the overloaded functions ODCITablePrepare and ODCITableDescribe in exceltableimpl, I can get it to compile.  Do you know anyway that I can pass the sheet_name into the cursor without running into this issue?

 

Thank you,

 

Mike

**Comment by ct3inc** [comment]
What version of Oracle are you using?  I’m still not able to get those files to give me any rows returned.  I’m using Oracle 18.5.

 

begin

  ExcelTable.useSheetPattern(true);

end;

/

 

SQL> 

  1  select *

  2  from table(

  3  ExcelTable.getRows(

  4  ExcelTable.getFile('TEST_DIR','SB_GDRC5-6-19.xlsx')

  5  , '.*'

  6  , q'{

  7  "C1"          varchar2(50)

  8  , "SHEET_NAME"  varchar2(31 char) for metadata (sheet_name)

  9  }'

10  )

11  ) x

12*

SQL> /

 

no rows selected

 

Thank you,

 

Mike

**Comment by mbleron** [comment]
I've tested the SQL queries on 12.1 and 18.3.
Everything's working OK, so no idea what's wrong in your case.

**Regarding the compilation issue** : 
That's because only static (literal) values are taken into account when Oracle calls ODCITableDescribe at compile time.
Since p_sheet_name is a bind variable, Oracle passes an untyped NULL value and therefore doesn't know if it's a sheet (VARCHAR2) or sheet list (ExcelTableSheetList) and fails to dispatch to the correct overload.
Named parameters are apparently ignored in this context.

The immediate workaround is to pass the sheet name as a constant string, if it's acceptable for you.
Otherwise, keep disabling the overloads in ExcelTableImpl.

The ExcelTable.getCursor procedure (which was created as a workaround in versions that don't support ANYDATASET-based static cursor) has the same drawback.
I need some time to work on that.

**Comment by ct3inc** [comment]
I think I figure out a work around.  It appears that an xls file works while the xlsx doesn’t.  

 

SQL> l

  1  select *

  2  from table(

  3  ExcelTable.getRows(

  4  ExcelTable.getFile('TEST_DIR','SB_GDRC 5-6-19.xls')

  5  , '.*'

  6  , q'{

  7  "C1"          varchar2(50)

  8  , "SHEET_NAME"  varchar2(31 char) for metadata (sheet_name)

  9  }',

10  '1:10'

11  )

12  ) x

13*

SQL> /

 

C1         SHEET_NAME

---------- --------------------

SBC #      GDRC

230389     GDRC

230377     GDRC

SBC #      GDRC BOX WOOD

 

SQL> l

  1  select *

  2  from table(

  3  ExcelTable.getRows(

  4  ExcelTable.getFile('TEST_DIR','SB_GDRC5-6-19.xlsx')

  5  , '.*'

  6  , q'{

  7  "C1"          varchar2(50)

  8  , "SHEET_NAME"  varchar2(31 char) for metadata (sheet_name)

  9  }',

10  '1:10'

11  )

12  ) x

13*

SQL> /

 

no rows selected

 

If I select more columns for the xlsx files, then I’m able to get results for the sheet_name:

  1  select *

  2  from table(

  3  ExcelTable.getRows(

  4  ExcelTable.getFile('TEST_DIR','SB_GDRC5-6-19.xlsx')

  5  , '.*'

  6  , q'{

  7  "C1"          varchar2(50),

  8  "C2"          varchar2(50),

  9  "C3"          varchar2(50),

10  "C4"          varchar2(50),

11  "C5"          varchar2(50),

12  "C6"          varchar2(50),

13  "C7"          varchar2(50),

14  "C8"          varchar2(50),

15  "C9"          varchar2(50),

16  "C10"          varchar2(50)

17  , "SHEET_NAME"  varchar2(31 char) for metadata (sheet_name)

18  }',

19  '1:10'

20  )

21  ) x

22*

SQL> /

 

C1         C2         C3         C4         C5         C6         C7

---------- ---------- ---------- ---------- ---------- ---------- ----------

C8         C9         C10        SHEET_NAME

---------- ---------- ---------- --------------------

                      GDRC LOGS

                       5/6/19

                                 GDRC

 

 

                                 GDRC

 

 

                                 GDRC

 

SBC #      LOAD       DATE       OWNER      TICKET #   SPECIE     NET

RATE       $/LOAD                GDRC

 

230389     17         43591      GDRC       9514618-19 FIR        64200

44         1412.4                GDRC

 

230377     29         43591      GDRC       9514608-09 FIR        62140

44         1367.08               GDRC

 

                      GDRC BOX W

                     OOD  9/28/

                      17

                                 GDRC BOX WOOD

 

 

                                 GDRC BOX WOOD

 

 

                                 GDRC BOX WOOD

 

SBC #      LOAD       DATE       OWNER      TICKET #   SPECIE     NET

RATE       $/LOAD                GDRC BOX WOOD

 

                                                                  #REF!

           #REF!                 GDRC BOX WOOD

 

 

                                 GDRC BOX WOOD

 

 

12 rows selected.

 

Not sure why that work this way but at least I can use your program.  

 

Thank you so much with your ehlp and time on this!

 

Mike

**Comment by mbleron** [comment]
Thanks for your last feedback Mike.

I've been investigating a bit further.
The empty result turns out to be a bug actually.
It occurs when the client asks only a few rows at a time, such as SQLPLUS which first requests the database to fetch a single row, or while fetching row-by-row from a PL/SQL cursor.
Basically, I was decrementing the count of fetched rows even though the row was empty (i.e. not containing any of the requested columns).

In the present case, for file SB_GDRC.5-6-19.xlsx, the first non-empty row in the range (1:10) is row 3, but the first non-empty column is C, so if we request only A (or A and B), ExcelTable will return an empty rowset and stop processing the file : 

_(trace added for debugging purpose)_

First two columns : 
```
SQL> set serveroutput on
SQL> select *
from table(
       ExcelTable.getRows(
         ExcelTable.getFile('PUBLIC_DIR','SB_GDRC.5-6-19.xlsx')
       , '.*'
       , '"C1" varchar2(15), "C2" varchar2(15)'
       , '1:10'
       )
     )
;

no rows selected

Requested rows = 1
```

Adding the third column : 
```
SQL> select *
from table(
       ExcelTable.getRows(
         ExcelTable.getFile('PUBLIC_DIR','SB_GDRC.5-6-19.xlsx')
       , '.*'
       , '"C1" varchar2(15), "C2" varchar2(15), "C3" varchar2(15)'
       , '1:10'
       )
     )
;

C1              C2              C3
--------------- --------------- ---------------
                                GDRC LOGS  5/6/


SBC #           LOAD            DATE
230389          17              43591
230377          29              43591
                                GDRC BOX WOOD


SBC #           LOAD            DATE


11 rows selected.

Requested rows = 1
Requested rows = 15
Requested rows = 5
```

I've published a fix in v3.2.1.

**Comment by ct3inc** [comment]
Thank you so much!  Your program is excellent and has really help me saved me a lot of time.

 

Mike
